subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c

```diff
@@ -0,0 +1,1217 @@
+		if (job_id != saved_job.id) {
+			strncpy(saved_job.id, job_id, sizeof(saved_job.id));
+		}
+	}
+
+	ret = settings_save_one(SETTINGS_FULL_FOTA_JOB, &saved_job,
+				sizeof(saved_job));
+	if (ret) {
+		LOG_ERR("settings_save_one failed: %d", ret);
+	}
+
+	return ret;
+}
+
+static void http_fota_handler(const struct fota_download_evt *evt)
+{
+	__ASSERT_NO_MSG(evt != NULL);
+
+	LOG_DBG("evt: %d", evt->id);
+
+	switch (evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		/* MCUBOOT: download finished, update job status and reboot */
+		current_fota.status = NRF_CLOUD_FOTA_IN_PROGRESS;
+		save_validate_status(current_fota.info.id,
+				     current_fota.info.type,
+				     NRF_CLOUD_FOTA_VALIDATE_PENDING);
+		send_job_update(&current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		/* MODEM: update job status and reboot */
+		current_fota.status = NRF_CLOUD_FOTA_IN_PROGRESS;
+		save_validate_status(current_fota.info.id,
+				     current_fota.info.type,
+				     NRF_CLOUD_FOTA_VALIDATE_PENDING);
+		send_job_update(&current_fota);
+		send_event(NRF_CLOUD_FOTA_EVT_ERASE_PENDING, &current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		/* MODEM: this event is received when the initial
+		 * fragment is downloaded and dfu_target_modem_init() is
+		 * called.
+		 */
+		send_event(NRF_CLOUD_FOTA_EVT_ERASE_DONE, &current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		if (last_fota_dl_evt == FOTA_DOWNLOAD_EVT_ERASE_DONE) {
+			current_fota.status = NRF_CLOUD_FOTA_REJECTED;
+		} else {
+			current_fota.status = NRF_CLOUD_FOTA_FAILED;
+			current_fota.error = NRF_CLOUD_FOTA_ERROR_DOWNLOAD;
+		}
+		save_validate_status(current_fota.info.id,
+				     current_fota.info.type,
+				     NRF_CLOUD_FOTA_VALIDATE_DONE);
+		send_job_update(&current_fota);
+		send_event(NRF_CLOUD_FOTA_EVT_ERROR, &current_fota);
+		cleanup_job(&current_fota);
+		break;
+
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		current_fota.status = NRF_CLOUD_FOTA_DOWNLOADING;
+		current_fota.dl_progress = evt->progress;
+
+		/* Do not send complete status more than once */
+		if ((current_fota.sent_dl_progress == 100) &&
+		    (current_fota.dl_progress == 100)) {
+			break;
+		}
+
+		/* Reset if new progress is less than previous */
+		if (current_fota.sent_dl_progress >
+		    current_fota.dl_progress) {
+			current_fota.sent_dl_progress = 0;
+		}
+
+		/* Only send progress update when finished and if increment
+		 * threshold is met
+		 */
+		if (current_fota.dl_progress != 100 &&
+		    ((current_fota.dl_progress -
+		      current_fota.sent_dl_progress) <
+		     CONFIG_NRF_CLOUD_FOTA_PROGRESS_PCT_INCREMENT)) {
+			break;
+		}
+
+		current_fota.sent_dl_progress = current_fota.dl_progress;
+		send_job_update(&current_fota);
+		break;
+	default:
+		break;
+	}
+
+	last_fota_dl_evt = evt->id;
+}
+
+static bool add_string_to_array(cJSON *const array, const char *const string)
+{
+	cJSON *item = cJSON_CreateString(string);
+
+	if (item && array) {
```

Comment:
If `array` is NULL we have a memory leak (no one frees `item`). 

I know it's a static function and it's likely not expected to provide a NULL `array` pointer, but it would be more readable to verify it serparately (with an assert for instance).